library/std/src/io/mod.rs

```diff
@@ -2777,23 +2777,50 @@ pub struct Bytes<R> {
 impl<R: Read> Iterator for Bytes<R> {
     type Item = Result<u8>;
 
-    #[inline]
+    // Not `#[inline]`. This function gets inlined even without it, but having
+    // the inline annotation can result in worse code generation. See #116785.
     fn next(&mut self) -> Option<Result<u8>> {
+        SpecReadByte::spec_read_byte(&mut self.inner)
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        SizeHint::size_hint(&self.inner)
+    }
+}
+
+// Trait for specialization of `Bytes::next`.
+trait SpecReadByte {
+    // Specializable method for reading a single byte.
+    fn spec_read_byte(&mut self) -> Option<Result<u8>>;
+
+    // Non-specializable method for reading a single byte. This is used by the
+    // default `spec_read_byte` and can also be called on a slow path of a
+    // specialized `spec_read_byte` method.
+    fn slow_read_byte(&mut self) -> Option<Result<u8>>;
+}
+
+impl<R> SpecReadByte for R
+where
+    Self: Read,
+{
+    #[inline]
+    default fn spec_read_byte(&mut self) -> Option<Result<u8>> {
+        self.slow_read_byte()
+    }
+
+    #[inline(never)]
```

Comment:
This now prevents inlining for i.e. non-BufReaders. Might be a bad idea for reading bytes from a `&mut &[u8]` or `VecDeque<u8>`. Is that intentional?
If not you could use a free `fn default_byte_read` and have the default `spec_read_byte` call that?

Do these even need to be trait functions at all? Two free functions (normal, non-inlinable) should also work.

Reply:
yeah the spec one needs to be in a trait. I meant a `read_normal` and `read_noinline` with the `default fn spec_read_byte` caulling `read_normal` while the `BufReader` one calls `read_noinline`.

Reply:
:+1: